mne/io/egi/egimff.py

```diff
@@ -0,0 +1,368 @@
+            with open(file_bin, 'rb') as fid:
+                block_info = _block_r(fid)
+            egi_info['block_info'] = block_info
+            egi_info['egi_events'] = egi_events
+
+            self._filenames = [file_bin]
+            self._raw_extras = [egi_info]
+        else:
+            raise ValueError("kind must be 'raw' or 'epoch'. Got %s." % kind)
+
+        super(RawMff, self).__init__(
+            info, preload=preload, orig_format=egi_info['orig_format'],
+            filenames=[file_bin], last_samps=[egi_info['n_samples'] - 1],
+            raw_extras=[egi_info], verbose=verbose)
+
+    def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
+        """Read a chunk of data."""
+        from ..utils import _mult_cal_one
+        egi_info = self._raw_extras[fi]
+        block_info = egi_info['block_info']
+        offset = block_info['position'] - 4
+        n_channels = egi_info['n_channels']
+        block_size = block_info['hl']
+        dtype = '<f4'
+        n_bytes = np.dtype(dtype).itemsize
+
+        data_offset = n_channels * start * n_bytes + offset
+        data_left = (stop - start) * n_channels
+
+        egi_events = egi_info['egi_events'][:, start:stop]
+        extra_samps = (start / block_info['nsamples'])
+        beginning = extra_samps * block_size
+        data_left += (data_offset - offset) / n_bytes - beginning
+        with open(self._filenames[fi], 'rb', buffering=0) as fid:
+            fid.seek(beginning * n_bytes + offset + extra_samps * n_bytes)
+            # extract data in chunks
+            sample_start = 0
+            # s_offset determines the offset inside the block in samples.
+            s_offset = (data_offset / n_bytes - beginning) / n_channels
+            while sample_start * n_channels < data_left:
+                fid.seek(4, 1)
+                block = np.fromfile(fid, dtype, block_size)
+                block = block.reshape(n_channels, -1, order='C')
+                count = data_left - sample_start * n_channels
+                end = count / n_channels
+                if sample_start == 0:
+                    block = block[:, s_offset:end]
+                    sample_sl = slice(sample_start,
+                                      sample_start + block.shape[1])
+                elif count < block_size:
+                    block = block[:, :end]
+                if sample_start != 0:
+                    sample_sl = slice(sample_start - s_offset,
+                                      sample_start - s_offset + block.shape[1])
+                data_view = data[:n_channels, sample_sl]
+                sample_start = sample_start + block_info['nsamples']
+                _mult_cal_one(data_view, block, idx, cals[:n_channels], mult)
+
+        # STI 014 is simply the sum of all event channels (powers of 2).
+        if len(egi_events) > 0:
+            egi_events = np.vstack([egi_events, np.sum(egi_events, axis=0)])
+            data[n_channels:] = egi_events
```

Comment:
I also made the stim channel construction lazy like this. Now it's created on data fetch stage.